dartagnan/src/main/java/com/dat3m/dartagnan/verification/solving/RefinementSolver.java

```diff
@@ -338,31 +337,21 @@ private static Relation getCopyOfRelation(Relation rel, Wmm m) {
-        if (r instanceof RelUnion) {
-            return new RelUnion(first, second);
-        } else if (r instanceof RelIntersection) {
-            return new RelIntersection(first, second);
-        } else if (r instanceof RelMinus) {
-            return new RelMinus(first, second);
-        } else if (r instanceof RelComposition) {
-            return new RelComposition(first, second);
-        } else if (r instanceof RelInverse) {
-            return new RelInverse(first);
-        } else if (r instanceof RelDomainIdentity) {
-            return new RelDomainIdentity(first);
-        } else if (r instanceof RelRangeIdentity) {
-            return new RelRangeIdentity(first);
-        } else if (r instanceof RelTrans) {
-            return new RelTrans(first);
-        } else if (r instanceof RelSetIdentity) {
-            return new RelSetIdentity(((RelSetIdentity) r).getFilter());
-        } else if (r instanceof RelCartesian) {
-            return new RelCartesian(((RelCartesian) r).getFirstFilter(), ((RelCartesian) r).getSecondFilter());
-        } else if (r instanceof RelFencerel) {
-            return new RelFencerel(((RelFencerel) r).getFilter());
-        }
-        throw new UnsupportedOperationException("copying relation " + r);
+    private static final class RelationCopier implements Relation.Visitor<Relation> {
+        final Wmm targetModel;
+        RelationCopier(Wmm m) { targetModel = m; }
+        @Override public Relation visitUnion(Relation rel, Relation... r) { return new RelUnion(copy(r[0]), copy(r[1])); }
+        @Override public Relation visitIntersection(Relation rel, Relation... r) { return new RelIntersection(copy(r[0]), copy(r[1])); }
+        @Override public Relation visitDifference(Relation rel, Relation r1, Relation r2) { return new RelMinus(copy(r1), copy(r2)); }
+        @Override public Relation visitComposition(Relation rel, Relation r1, Relation r2) { return new RelComposition(copy(r1), copy(r2)); }
+        @Override public Relation visitInverse(Relation rel, Relation r1) { return new RelInverse(copy(r1)); }
+        @Override public Relation visitDomainIdentity(Relation rel, Relation r1) { return new RelDomainIdentity(copy(r1)); }
+        @Override public Relation visitRangeIdentity(Relation rel, Relation r1) { return new RelRangeIdentity(copy(r1)); }
+        @Override public Relation visitTransitiveClosure(Relation rel, Relation r1) { return new RelTrans(copy(r1)); }
+        @Override public Relation visitIdentity(Relation rel, FilterAbstract filter) { return new RelSetIdentity(filter); }
+        @Override public Relation visitProduct(Relation rel, FilterAbstract f1, FilterAbstract f2) { return new RelCartesian(f1, f2); }
+        @Override public Relation visitFences(Relation rel, FilterAbstract type) { return new RelFencerel(type); }
+        private Relation copy(Relation r) { return getCopyOfRelation(r, targetModel); }
```

Comment:
Does this somehow interfere with with the PR from @ThomasHaas for solving the problem related to cutting?

Reply:
It should not be related, other than refactoring some part two lines before that.

Reply:
Assuming you referred to #325.

Reply:
Automatic merging could have missed this.  However, the visitor still does not add the copy to the model by itself.

(Not exactly refactoring: Trying to copy an unnamed relation that has neither of the listed definitions (i.e. `RecursiveRelation`) will now throw a `NullPointer` instead of `UnsupportedOperation`.)

Reply:
I think the visitor will simply create a new instance of `Relation` and then the fix in #325 will take care of adding the relation to the memory model. I just wanted to double check that this would not again cause an issue that the CI did not noticed (because we don't have much tests for cutting).

I do not get your last comment about throwing a `NullPointer` exception instead of a `UnsupportedOperation` one.

Reply:
The fallback branch in the replaced code threw an `UnsupportedOperationException`.  `Relation.Visitor` falls back to returning `null`.  In that case, nothing is checked (yet), so some parent relations will contain null references in the copied model.

Reply:
Instead of checking if `copy` returned null, wouldn't it make more sense to override `visitDefinition` in `RelationCopier` and throw the exception here?

Reply:
That would be even better